File: Python/Partition List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    def partition(self, head, x):
        head1 = None
        head2 = None
        node1 = None
        node2 = None
        node = head
        while node != None:
            if node.val < x:
                if head1 == None:
                    head1 = node
                    node1 = node
                else:
                    node1.next = node
                    node1 = node1.next
            else:
                if head2 == None:
                    head2 = node
                    node2 = node
                else:
                    node2.next = node
                    node2 = node2.next
            node = node.next
        if head1 != None:
            node1.next = head2
        if node2 != None:
            node2.next = None
        if head1 != None:
            return head1
        else:
            return head2

File: C++/Partition-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *dummy1 = new ListNode(0), *dummy2 = new ListNode(0);
        ListNode *leftcur = dummy1, *rightcur = dummy2;
        while (head != NULL) {
            if (head->val < x) {
                leftcur->next = head;
                leftcur = leftcur->next;
            } else {
                rightcur->next = head;
                rightcur = rightcur->next;
            }
            head = head->next;
        }
        rightcur->next = NULL;
        leftcur->next = dummy2->next;
        return dummy1->next;
    }
};